bpo-32270: Don't close stdin/out/err in pass_fds

gpshead opened the pull request:
When subprocess.Popen() stdin= stdout= or stderr= handles are specified and
appear in pass_fds=, don't close their original fds after dup2'ing them to 0/1/2.

I still need to create a proper unittest and NEWS entry before this can be merged.

<!-- issue-number: bpo-32270 -->
https://bugs.python.org/issue32270
<!-- /issue-number -->

gpshead committed:
bpo-32270: Don't close stdin/out/err in pass_fds

When subprocess.Popen() stdin= stdout= or stderr= handles are specified
and appear in pass_fds=, don't close the original fds after dup'ing them.

izbyshev commented on Modules/_posixsubprocess.c:
```diff
@@ -453,11 +453,13 @@ child_exec(char *const exec_array[],
     /* Close pipe fds.  Make sure we don't close the same fd more than */
     /* once, or standard fds. */
-    if (p2cread > 2)
+    if (p2cread > 2 && !_is_fd_in_sorted_fd_sequence(p2cread, py_fds_to_keep))
```
Note that there is no need to manually close `p2cread` at all, and other two descriptors need to be closed only if they are the result of `dup` above. This is because if a descriptor is created by subprocess.py (a pipe or for `/dev/null`), it's non-inheritable, and if a descriptor is specified by the user, is not in `pass_fds` and is > 2, it will be closed by `_close_open_fds`. So my original idea of fixing this issue was to add an async-signal-safe version of `_Py_dup` and use it instead of `dup` above, and then just remove these three checks.

Note also that `p2cwrite` and its three friends are never required to be manually closed since they are always non-inheritable pipe ends created by subprocess.py.

gpshead replied:
Not surprising. The original code here predates our non-inheritable file descriptor implementation. I'm examining this vs your https://github.com/izbyshev/cpython/commit/b89b52f28490b69142d5c061604b3a3989cec66c changes to see what makes the most sense.

This change as is _does_ make it pass your unittest.

gpshead committed:
Adds a unittest from @izbyshev.

Taken from https://github.com/izbyshev/cpython/commit/b89b52f28490b69142d5c061604b3a3989cec66c
per the suggestion in the BPO issue comment.

gpshead committed:
Set c2pwrite & errwrite non-inheritable when duped

Based on code review by izbyshev@ and
https://github.com/izbyshev/cpython/commit/b89b52f28490b69142d5c061604b3a3989cec66c.

This also removes the old manual p2cread, c2pwrite, and errwrite closing logic
as inheritable flags and _close_open_fds takes care of them properly today.

This code is within child_exec() where it is the only thread so there is no
race condition between the dup and _Py_set_inheritable_async_safe call.

gpshead committed:
NEWS!
